### PR TITLE
Lib cleanup

### PR DESCRIPTION
**Pull request by jdalsem**

### CODE REVIEW

File: docs/appendix/upgrade-notes/3.x-to-4.0.rst [hunk 1]
```diff
@@ -53,6 +53,12 @@ complicated validating responses (for example in hooks). The classes have been s
 
 All classes now extend ``Elgg\Http\Response`` and implement ``Elgg\Http\ReponseBuilder``.
 
+Default widgets
+---------------
+
+The magic handling the creation of default widgets has been reduced. You now need to register the ``Elgg\Widgets\CreateDefaultWidgetsHandler`` callback to the event when you want default widgets to be created.
```
Comment (jeabakker) [inline docs/appendix/upgrade-notes/3.x-to-4.0.rst]: > The magic handling the creation

The magic handling **for** the creation

File: docs/guides/widgets.rst [hunk 2]
```diff
@@ -190,6 +190,20 @@ To announce default widget support in your plugin, register for the ``get_list,
 .. code-block:: php
 
     elgg_register_plugin_hook_handler('get_list', 'default_widgets', 'my_plugin_default_widgets_hook');
+    
+    function my_plugin_default_widgets_hook($hook, $type, $return, $params) {
+        $return[] = array(
```
Comment (jeabakker) [inline docs/guides/widgets.rst]: short array notation please (as per the coding standards)

File: engine/classes/Elgg/Notifications/ProcessQueueCronHandler.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	 * Process notification queue
+	 *
+	 * @param \Elgg\Hook $hook 'cron', 'minute'
+	 *
+	 * @return void
+	 */
+	public function __invoke(\Elgg\Hook $hook) {
+		// calculate when we should stop
+		// @todo make configurable?
+		$stop_time = time() + 45;
```
Comment (jeabakker) [inline engine/classes/Elgg/Notifications/ProcessQueueCronHandler.php]: (base) time should come the hook eg ``$hook->getParam('time')``

File: engine/classes/Elgg/Notifications/SendEmailHandler.php [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+			return;
+		}
+	
+		$message = $hook->getParam('notification');
+		if (!$message instanceof \Elgg\Notifications\Notification) {
+			return false;
+		}
+	
+		$sender = $message->getSender();
+		$recipient = $message->getRecipient();
```
Comment (jeabakker) [inline engine/classes/Elgg/Notifications/SendEmailHandler.php]: move ``$recipient`` to just before the checks

File: engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		static $processed_events = [];
+		
+		if (isset($processed_events["{$event->getName()}.{$event->getType()}"])) {
+			return;
+		}
+		// only create default widgets once per event
+		$processed_events["{$event->getName()}.{$event->getType()}"] = true;
+		
+		$default_widget_info = elgg_trigger_plugin_hook('get_list', 'default_widgets', null, []);
+		$entity = $event->getObject();
```
Comment (jeabakker) [inline engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php]: you can save a plugin hook trigger if you first check for a valid entity

File: engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+			elgg_push_context('create_default_widgets');
+	
+			elgg_call(ELGG_IGNORE_ACCESS, function () use ($entity, $info) {
+				// pull in by widget context with widget owners as the site
+				// not using elgg_get_widgets() because it sorts by columns and we don't care right now.
+				$widgets = elgg_get_entities([
+					'type' => 'object',
+					'subtype' => 'widget',
+					'owner_guid' => elgg_get_site_entity()->guid,
+					'private_setting_name' => 'context',
```
Comment (jeabakker) [inline engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php]: why not a ``private_setting_name_value_pair``?

File: engine/classes/Elgg/Widgets/DefaultWidgetsContainerPermissionsHandler.php [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+	
+	/**
+	 * Overrides permissions checks when creating widgets for logged out users.
+	 *
+	 * @param \Elgg\Hook $hook 'container_permissions_check', 'object'
+	 *
+	 * @return void|true
+	 */
+	public function __invoke(\Elgg\Hook $hook) {
+		if ($hook->getType() !== 'object' || $hook->getParam('subtype') !== 'widget') {
```
Comment (jeabakker) [inline engine/classes/Elgg/Widgets/DefaultWidgetsContainerPermissionsHandler.php]: the hook is registered for type ``object`` so why check again?

File: engine/hooks.php [hunk 8]
```diff
@@ -161,11 +158,13 @@
 		'menu:page' => [
 			'Elgg\Menus\Page::registerAdminAdminister' => [],
 			'Elgg\Menus\Page::registerAdminConfigure' => [],
+			'Elgg\Menus\Page::registerAdminDefaultWidgets' => [],
 			'Elgg\Menus\Page::registerAdminInformation' => [],
 			'Elgg\Menus\Page::registerAdminPluginSettings' => [],
 			'Elgg\Menus\Page::registerAvatarEdit' => [],
 			'Elgg\Menus\Page::registerUserSettings' => [],
 			'Elgg\Menus\Page::registerUserSettingsPlugins' => [],
+
```
Comment (jeabakker) [inline engine/hooks.php]: oops

**Review by jeabakker — CHANGES_REQUESTED** [review]

File: docs/appendix/upgrade-notes/3.x-to-4.0.rst (hunk 1 above)
Comment (jdalsem) [reply docs/appendix/upgrade-notes/3.x-to-4.0.rst]: nope

File: engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php (hunk 6 above)
Comment (jdalsem) [reply engine/classes/Elgg/Widgets/CreateDefaultWidgetsHandler.php]: jjust a copy and paste

File: engine/classes/Elgg/Widgets/DefaultWidgetsContainerPermissionsHandler.php (hunk 7 above)
Comment (jdalsem) [reply engine/classes/Elgg/Widgets/DefaultWidgetsContainerPermissionsHandler.php]: no reason.. will change

**Review by jeabakker — APPROVED** [review]

**Pull request merged**